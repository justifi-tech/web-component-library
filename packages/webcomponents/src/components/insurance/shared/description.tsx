import { Component, h } from '@stencil/core';

@Component({
  tag: 'justifi-insurance-description',
  shadow: true,
})

export class InsuranceDescription {
  render() {
    <div></div>
  };
};