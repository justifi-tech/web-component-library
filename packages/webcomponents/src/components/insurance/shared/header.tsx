import { Component, h } from '@stencil/core';

@Component({
  tag: 'justifi-insurance-header',
  shadow: true,
})

export class InsuranceHeader {
  render() {
    <div></div>
  };
};