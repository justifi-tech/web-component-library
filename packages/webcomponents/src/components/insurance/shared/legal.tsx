import { Component, h } from '@stencil/core';

@Component({
  tag: 'justifi-insurance-legal',
  shadow: true,
})

export class InsuranceLegal {
  render() {
    <div>
      <slot></slot>
    </div>
  };
};