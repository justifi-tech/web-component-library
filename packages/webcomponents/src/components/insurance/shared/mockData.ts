export const quote = {
  "id": "4b154cc4-6f1a-47b3-bdd9-3d22c9027b24",
  "master_policy_id": "4b154cc4-6f1a-47b3-bdd9-3d22c9027b24",
  "partner": {
    "id": "a21c59ee-2fd9-45c0-b6b3-8402bc7c67ba",
    "legal_business_name": "Vertical Insure, Inc."
  },
  "policy_status": "QUOTED",
  "expiration_date": "2024-02-29",
  "effective_date": "2024-02-28",
  "quote_date": "2024-02-28T19:41:02.840256Z",
  "product": {
    "id": "716f9506-aeb1-4738-9b1a-6e48e90dc0f6",
    "name": "registration-cancellation",
    "friendly_name": "Registration Cancellation Insurance",
    "abbreviation": "RC",
    "description": "Spinnaker Registration Cancellation - No Work Coverage",
    "perils": [
      "Illness (including COVID-19)",
      "Injury",
      "Weather / Natural Disasters",
      "Employment",
      "Transportation",
      "Many other covered reasons"
    ],
    "promotional_header": "Add Registration Cancellation Protection",
    "promotional_description": "<p>Protect your registration and get your money back in the event of unforeseen circumstances that prevent participation, including sickness (including COVID-19), injury, inclement weather, transportation, job loss and other covered reasons.</p>",
    "legal_disclaimer": "<p>battleface Travel Insurance plans are underwritten by Spinnaker Insurance Company (an IL Corporation, NAIC# 24376), with administrative office at One Pluckemin Way, Suite 102, Bedminster, NJ 07921. Plans are offered and administered by battleface Insurance Services LLC, 45 East Lincoln Street, Columbus, OH 43215, National Producer Number 18731960 (FL License number L107363/CA License number 0M75381). <a rel=\"noopener noreferrer\" href=\"https://verticalinsure.com/s/Vertical-Insure-Disclosures-SPI-040123.pdf\" target=\"_blank\">Terms and conditions apply</a>. <a rel=\"noopener noreferrer\" href=\"https://www.battleface.com/en-us/spi-vertical-insure-library/\" target=\"_blank\">You can view the policy here</a>.</p>"
  },
  "policy_attributes": {
    "@type": "registration-cancellation",
    "carrier_product_code": "1010",
    "coverage_date": "DAY_OF",
    "participant": {
      "first_name": "John",
      "last_name": "Doe"
    },
    "insurable_amount": 50000,
    "event_start_date": "2024-02-29",
    "event_end_date": "2024-02-29"
  },
  "premium_amount": 3375,
  "currency": "USD",
  "total": 3375,
  "quote_id": "r1azwq4l0u9dzwnpd7dtu48t",
  "is_test": true,
  "checkout_url": "https://landing.verticalinsure.com/verticalinsure/registration-cancellation/r1azwq4l0u9dzwnpd7dtu48t?id=eyJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJ0ZXN0MDFAdmVydGljYWxpbnN1cmUuY29tIiwiZW52aXJvbm1lbnQiOiJURVNUIiwicmVzb3VyY2UiOiJyMWF6d3E0bDB1OWR6d25wZDdkdHU0OHQiLCJwZXJtaXNzaW9ucyI6WyJQVVJDSEFTRV9QT0xJQ1kiXSwiaXNzIjoiaHR0cHM6Ly92ZXJ0aWNhbGluc3VyZS5jb20iLCJleHAiOjE3MTAzNTk3MzUsImlhdCI6MTcwOTE1MDEzNX0.ICLtb6MLQ9LuNOZi_2XYiPpt3ggiUoN-nYbEchL7ZpA",
  "policy_holder": {
    "email_address": "test01@verticalinsure.com",
    "country": "US",
    "id": "20c1d250-89d0-3904-84fc-5ee46cf2cee3",
    "state": "MN"
  }
};
