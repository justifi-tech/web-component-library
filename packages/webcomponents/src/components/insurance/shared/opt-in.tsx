import { Component, h } from '@stencil/core';

@Component({
  tag: 'justifi-insurance-opt-in',
  shadow: true,
})

export class InsuranceOptIn {
  render() {
    <div></div>
  };
};