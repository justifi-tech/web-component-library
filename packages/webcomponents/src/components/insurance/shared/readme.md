# justifi-insurance-opt-in



<!-- Auto Generated Below -->


## Dependencies

### Used by

 - [justifi-insurance-registration-cancellation](../registration-cancellation)

### Graph
```mermaid
graph TD;
  justifi-insurance-registration-cancellation --> justifi-insurance-opt-in
  style justifi-insurance-opt-in fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
