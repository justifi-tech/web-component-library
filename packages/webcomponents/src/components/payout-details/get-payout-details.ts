import { Payout } from '../../api';
import { getErrorMessage } from '../../api/services/utils';

export const makeGetPayoutDetails =
  ({ id, authToken, service }) =>
  async ({ onSuccess, onError }) => {
    try {
      const response = await service.fetchPayout(id, authToken);

      if (!response.error) {
        const payout = new Payout(response.data);

        onSuccess(payout);
      } else {
        const responseError = getErrorMessage(response.error);
        const errorMessage = responseError;
        onError(getErrorMessage(errorMessage));
      }
    } catch (error) {
      const errorMessage = error;
      onError(getErrorMessage(errorMessage));
    }
  };
