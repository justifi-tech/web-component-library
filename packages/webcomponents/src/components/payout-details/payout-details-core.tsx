import { Component, h, Prop, State, Watch, Event, EventEmitter } from '@stencil/core';
import { IPayout, Payout } from '../../api';
import { formatCurrency, formatDate, formatTime } from '../../utils/utils';
import { CodeBlock, DetailItem, DetailSectionTitle, EntityHeadInfo, EntityHeadInfoItem, ErrorState, LoadingState } from '../../ui-components/details/utils';
import { ComponentError } from '../../api/ComponentError';
import { StyledHost } from '../../ui-components';
import { MapPayoutStatusToBadge } from '../payouts-list/payouts-status';

@Component({
  tag: 'payout-details-core',
})

export class PayoutDetailsCore {
  @Prop() getPayout: Function;
  @Prop() getPayoutCSV: Function;

  @State() payout: Payout;
  @State() loading: boolean = true;
  @State() errorMessage: string = null;

  @Event({ eventName: 'error-event' }) errorEvent: EventEmitter<ComponentError>;

  componentWillLoad() {
    if (this.getPayout) {
      this.fetchData();
    }
  }

  @Watch('getPayout')
  getPayoutWatcher() {
    this.fetchData();
  }

  fetchData(): void {
    this.loading = true;

    this.getPayout({
      onSuccess: (payout: IPayout) => {
        this.errorMessage = null;
        this.payout = payout;
        this.loading = false;
      },
      onError: ({ error, code, severity }) => {
        this.errorMessage = error;
        this.errorEvent.emit({
          message: error,
          errorCode: code,
          severity,
        });
        this.loading = false;
      },
    });
  }

  downloadCSV = () => {
    this.getPayoutCSV({
      payoutId: this.payout.id,
      onError: ({ error, code, severity }) => {
        this.errorMessage = error;
        this.errorEvent.emit({
          message: error,
          errorCode: code,
          severity,
        });
      },
    });
  }

  render() {
    return (
      <StyledHost>
        {this.loading && LoadingState()}
        {!this.loading && this.errorMessage && ErrorState(this.errorMessage)}
        {!this.loading && this.payout && (
          <justifi-details error-message={this.errorMessage}>
            <EntityHeadInfo
              slot="head-info"
              badge={<span slot='badge' part="detail-head-badge" innerHTML={MapPayoutStatusToBadge(this.payout?.status)} />}
              title={`${formatCurrency(this.payout.amount)} ${this.payout.currency.toUpperCase()}`}
            >
              <EntityHeadInfoItem
                classes="border-1 border-end"
                title="Updated At"
                value={`${formatDate(this.payout.updated_at)} ${formatTime(this.payout.updated_at)}`}
              />
              <EntityHeadInfoItem
                classes="border-1 border-end"
                title="Created At"
                value={`${formatDate(this.payout.created_at)} ${formatTime(this.payout.created_at)}`}
              />
              <EntityHeadInfoItem title="ID" value={this.payout.id} />
              <div class="m-4">
                <button part="button-secondary" class="btn btn-outline-secondary d-flex align-items-center" onClick={this.downloadCSV}>
                  Export CSV
                </button>
              </div>
            </EntityHeadInfo>
            <div slot='detail-sections' part="detail-section">
              <DetailSectionTitle sectionTitle="Details" />
              <div class="d-table gap-2 w-100">
                <DetailItem title="Date paid" value={formatDate(this.payout.deposits_at)} />
                <DetailItem title="Statement Description" value={this.payout.description} />
                <DetailItem title="Payout Method" value={this.payout.delivery_method} />
                <DetailItem title="Amount" value={formatCurrency(this.payout.amount)} />
                <DetailItem title="Fee" value={formatCurrency(this.payout.fees_total)} />
              </div>
              <DetailSectionTitle sectionTitle="Account" />
              <div class="d-table gap-2 w-100">
                <DetailItem title="ID" value={this.payout.account_id} />
                <DetailItem title="Account Type" value={this.payout.bank_account.account_type} />
                <DetailItem title="Institution" value={this.payout.bank_account.account_type} />
                <DetailItem title="Routing Number" value={this.payout.bank_account.routing_number} />
                <DetailItem title="Account Number" value={this.payout.bank_account.account_number_last4} />
              </div>
              <DetailSectionTitle sectionTitle='Metadata' />
              <div class="d-table gap-2 w-100">
                <CodeBlock metadata={this.payout.metadata} />
              </div>
            </div>
          </justifi-details>
        )}
      </StyledHost>
    )
  }
}
