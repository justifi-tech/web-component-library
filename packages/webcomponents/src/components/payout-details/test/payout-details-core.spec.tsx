jest.mock('../../../ui-components/styled-host/modified-bootstrap.css', () => '');

import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import { PayoutDetailsCore } from '../payout-details-core';
import mockPayoutDetailSuccess from '../../../../../../mockData/mockPayoutDetailsSuccess.json';
import { makeGetPayoutDetails } from '../get-payout-details';
import { API_NOT_AUTHENTICATED_ERROR } from '../../../api/shared';

describe('payout-details-core', () => {
  it('renders loading state initially', async () => {
    // Define a mock getPayout function as part of the component initialization
    const mockGetPayout = jest.fn().mockImplementation(({ onSuccess }) => {
      // Simulate an asynchronous operation without resolving immediately
      setTimeout(() => onSuccess({ /* mock data */ }), 1000);
    });

    // Initialize the page with the mock getPayout function included in the component tag
    const page = await newSpecPage({
      components: [PayoutDetailsCore],
      template: () => <payout-details-core getPayout={mockGetPayout} />
    });

    // Wait for any state changes to complete
    await page.waitForChanges();

    // Assert the loading state is visible in the snapshot
    expect(page.root).toMatchSnapshot();
  });

  it('fetches payout details on connected callback', async () => {
    const mockService = {
      fetchPayout: jest.fn().mockResolvedValue(mockPayoutDetailSuccess),
    };

    const getPayout = makeGetPayoutDetails({
      id: 'some-id',
      authToken: 'some-auth-token',
      service: mockService,
    });

    const page = await newSpecPage({
      components: [PayoutDetailsCore],
      template: () => <payout-details-core getPayout={getPayout} />
    });

    await page.waitForChanges();

    expect(mockService.fetchPayout).toHaveBeenCalled(); // Verify the mock service method was called
    expect(page.rootInstance.payout).toEqual(expect.objectContaining({
      "id": mockPayoutDetailSuccess.id, // Verify the component state was updated with mock data
    }));
    expect(page.root).toMatchSnapshot(); // Verify the rendered output
  });


  it('handles fetch payout error correctly', async () => {
    // Mock the service method that gets called by getPayout
    const mockService = {
      fetchPayout: jest.fn().mockRejectedValue(new Error('Fetch error'))
    };

    // Real getPayout function but with mocked service
    const getPayout = makeGetPayoutDetails({
      id: 'some-id', // Use appropriate id
      authToken: 'some-auth-token', // Use appropriate auth token
      service: mockService // Injecting the mocked service
    });

    const page = await newSpecPage({
      components: [PayoutDetailsCore],
      template: () => <payout-details-core getPayout={getPayout} />
    });

    await page.waitForChanges();

    // Assertions to verify the error handling
    expect(page.rootInstance.errorMessage).toBe('Fetch error');
    expect(page.root).toMatchSnapshot();
  });

  it('emits error event on fetch error', async () => {
    const mockService = {
      fetchPayout: jest.fn().mockRejectedValue(new Error('Fetch error'))
    };

    const getPayout = makeGetPayoutDetails({
      id: 'some-id',
      authToken: 'some-auth',
      service: mockService
    });

    const onErrorEvent = jest.fn();

    const page = await newSpecPage({
      components: [PayoutDetailsCore],
      template: () => <payout-details-core getPayout={getPayout} onError-event={onErrorEvent} />
    });

    await page.waitForChanges();

    expect(onErrorEvent).toHaveBeenCalledWith(expect.objectContaining({
      detail: {
        message: 'Fetch error',
        errorCode: 'fetch-error',
        severity: 'error'
      }
    }));
  });

  it('emits error event when API returns error', async () => {
    const mockService = {
      fetchPayout: jest.fn().mockResolvedValue(API_NOT_AUTHENTICATED_ERROR)
    };

    const getPayout = makeGetPayoutDetails({
      id: 'some-id',
      authToken: 'some-auth',
      service: mockService
    });

    const onErrorEvent = jest.fn();

    const page = await newSpecPage({
      components: [PayoutDetailsCore],
      template: () => <payout-details-core getPayout={getPayout} onError-event={onErrorEvent} />
    });

    await page.waitForChanges();

    expect(onErrorEvent).toHaveBeenCalledWith(expect.objectContaining({
      detail: {
        message: 'Not Authenticated',
        errorCode: 'not-authenticated',
        severity: 'error'
      }
    }));
  });
});
