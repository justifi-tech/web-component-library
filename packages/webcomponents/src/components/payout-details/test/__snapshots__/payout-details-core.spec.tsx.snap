// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`payout-details-core fetches payout details on connected callback 1`] = `
<payout-details-core class="body" style="display: block;">
  <style></style>
  <justifi-details>
    <div class="p-2" part="detail-head" slot="head-info">
      <div class="align-items-center d-flex flex-row gap-2 mb-2">
        <h1 class="m-0" part="detail-head-amount">
          $220.15 USD
        </h1>
        <span part="detail-head-badge" slot="badge">
          <span class="badge bg-success" title="Successfully deposited into your bank account">
            Paid
          </span>
        </span>
      </div>
      <div class="align-items-top d-flex flex-row" part="detail-head-info">
        <div class="border-1 border-end d-flex flex-column" part="detail-head-info-item">
          <span class="border-1 border-bottom fw-bold pe-2 ps-2" part="detail-head-info-item-title">
            Updated At
          </span>
          <span class="pe-2 ps-2 pt-2" part="detail-head-info-item-data">
            Dec 28, 2023 1:27am
          </span>
        </div>
        <div class="border-1 border-end d-flex flex-column" part="detail-head-info-item">
          <span class="border-1 border-bottom fw-bold pe-2 ps-2" part="detail-head-info-item-title">
            Created At
          </span>
          <span class="pe-2 ps-2 pt-2" part="detail-head-info-item-data">
            Dec 28, 2023 1:27am
          </span>
        </div>
        <div class="d-flex flex-column undefined" part="detail-head-info-item">
          <span class="border-1 border-bottom fw-bold pe-2 ps-2" part="detail-head-info-item-title">
            ID
          </span>
          <span class="pe-2 ps-2 pt-2" part="detail-head-info-item-data">
            po_17745yESnHyEgWNeunmhmR
          </span>
        </div>
        <div class="m-4">
          <button class="align-items-center btn btn-outline-secondary d-flex" part="button-secondary">
            Export CSV
          </button>
        </div>
      </div>
    </div>
    <div part="detail-section" slot="detail-sections">
      <h2 class="border-bottom fs-5 mt-4 pb-3" part="header-2">
        Details
      </h2>
      <div class="d-table gap-2 w-100">
        <div class="d-table-row gap-2" part="detail-section-item">
          <span class="d-table-cell fw-bold px-2" part="detail-section-item-title">
            Date paid
          </span>
          <span class="d-table-cell flex-1 px-2 text-wrap" part="detail-section-item-data">
            Dec 28, 2023
          </span>
        </div>
        <div class="d-table-row gap-2" part="detail-section-item">
          <span class="d-table-cell fw-bold px-2" part="detail-section-item-title">
            Statement Description
          </span>
          <span class="d-table-cell flex-1 px-2 text-wrap" part="detail-section-item-data">
            Payout
          </span>
        </div>
        <div class="d-table-row gap-2" part="detail-section-item">
          <span class="d-table-cell fw-bold px-2" part="detail-section-item-title">
            Payout Method
          </span>
          <span class="d-table-cell flex-1 px-2 text-wrap" part="detail-section-item-data">
            standard
          </span>
        </div>
        <div class="d-table-row gap-2" part="detail-section-item">
          <span class="d-table-cell fw-bold px-2" part="detail-section-item-title">
            Amount
          </span>
          <span class="d-table-cell flex-1 px-2 text-wrap" part="detail-section-item-data">
            $220.15
          </span>
        </div>
        <div class="d-table-row gap-2" part="detail-section-item">
          <span class="d-table-cell fw-bold px-2" part="detail-section-item-title">
            Fee
          </span>
          <span class="d-table-cell flex-1 px-2 text-wrap" part="detail-section-item-data">
            $14.85
          </span>
        </div>
      </div>
      <h2 class="border-bottom fs-5 mt-4 pb-3" part="header-2">
        Account
      </h2>
      <div class="d-table gap-2 w-100">
        <div class="d-table-row gap-2" part="detail-section-item">
          <span class="d-table-cell fw-bold px-2" part="detail-section-item-title">
            ID
          </span>
          <span class="d-table-cell flex-1 px-2 text-wrap" part="detail-section-item-data">
            acc_5Et9iXrSSAZR2KSouQGAWi
          </span>
        </div>
        <div class="d-table-row gap-2" part="detail-section-item">
          <span class="d-table-cell fw-bold px-2" part="detail-section-item-title">
            Account Type
          </span>
          <span class="d-table-cell flex-1 px-2 text-wrap" part="detail-section-item-data">
            checking
          </span>
        </div>
        <div class="d-table-row gap-2" part="detail-section-item">
          <span class="d-table-cell fw-bold px-2" part="detail-section-item-title">
            Institution
          </span>
          <span class="d-table-cell flex-1 px-2 text-wrap" part="detail-section-item-data">
            checking
          </span>
        </div>
        <div class="d-table-row gap-2" part="detail-section-item">
          <span class="d-table-cell fw-bold px-2" part="detail-section-item-title">
            Routing Number
          </span>
          <span class="d-table-cell flex-1 px-2 text-wrap" part="detail-section-item-data">
            110000000
          </span>
        </div>
        <div class="d-table-row gap-2" part="detail-section-item">
          <span class="d-table-cell fw-bold px-2" part="detail-section-item-title">
            Account Number
          </span>
          <span class="d-table-cell flex-1 px-2 text-wrap" part="detail-section-item-data">
            6789
          </span>
        </div>
      </div>
      <h2 class="border-bottom fs-5 mt-4 pb-3" part="header-2">
        Metadata
      </h2>
      <div class="d-table gap-2 w-100"></div>
    </div>
  </justifi-details>
</payout-details-core>
`;

exports[`payout-details-core handles fetch payout error correctly 1`] = `
<payout-details-core class="body" style="display: block;">
  <style></style>
  <main class="d-flex justify-content-center p-4 text-center" part="detail-empty-state" style="font-size: 1.2rem;">
    <div>
      Fetch error
    </div>
  </main>
</payout-details-core>
`;

exports[`payout-details-core renders loading state initially 1`] = `
<payout-details-core class="body" style="display: block;">
  <style></style>
  <main class="d-flex justify-content-center p-4 text-center" part="detail-loading-state" style="font-size: 1.2rem;">
    <div class="spinner-border spinner-border-sm" part="detail-loading-spinner" role="status">
      <span class="visually-hidden">
        Loading...
      </span>
    </div>
  </main>
</payout-details-core>
`;
